### PR TITLE
Make module http framework free

### PR DESCRIPTION
**Pull request by trygve-lie**
This PR makes the Context http framework free making it possible to integrate the Context with multiple http frameworks such as Hapi, Fastify etc.

Previously Podium has been very tied to Express.js where each module in the request chain has been a middleware where properties from each module has been written to `res.locals.podium`. Internally multiple of Express.js only methods have also been used making Podium very tied to Express.js.

This implementation switches from using a middleware chain passing request and response objects around to a Promise chain where a HttpIncoming object we have controll over is passed on. In other words; each module in Podium does now have a `.process()` method which can be chained together making it possible for any other http framwork to be integrated.

All usage of Express.js only methods has also been replaced with generic methods which should work within any http framework.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,56 +18,31 @@ $ npm install @podium/context
...
-// transform the context into a usable object.
-// The context will be stored at: `res.locals.podium.context`
-app.use(Context.deserialize());
+    // Serialize the context into a object that can be passed on
+    // to be passed on as http headers on a http request
+    const headers = Context.serialize({}, incom.context);
 
-app.get('/', (req, res) => {
-    res.status(200).json(res.locals.podium.context);
+    [ ... snip ...]
```
Comment (stipsan) [inline README.md]: ✂️

File: README.md [hunk 2]
```diff
@@ -14,56 +18,31 @@ $ npm install @podium/context
...
-app.listen(8080);
-```
-
-Simple Podlet server recieving requests from the Layout server:
-
-```js
-const express = require('express');
-const Context = require('@podium/context');
+const server = http.createServer(async (req, res) => {
+    // Create a HttpIncomming object
```
Comment (alexanbj) [inline README.md]: HttpIncoming

File: README.md [hunk 3]
```diff
@@ -1,10 +1,14 @@
...
+Module to generate the context which is passed on requests from a Podium Layout server to a Podium Podlet server.
+
 [![Build Status](https://travis-ci.org/podium-lib/context.svg?branch=master)](https://travis-ci.org/podium-lib/context)
 [![Greenkeeper badge](https://badges.greenkeeper.io/podium-lib/context.svg)](https://greenkeeper.io/)
 [![Known Vulnerabilities](https://snyk.io/test/github/podium-lib/context/badge.svg)](https://snyk.io/test/github/podium-lib/context)
 
-Module to generate, serialize and de-serialize a Podium Context.
+This module is intended for internal use in Podium and is not a module an end user would
+use directly. An end user can though interact with this module through a higher level
+module such as the @podium/layout module.
```
Comment (alexanbj) [inline README.md]: Make `@podium/layout` a link to either the package or repository

File: README.md [hunk 4]
```diff
@@ -192,47 +171,52 @@ This method takes the following arguments:
...
-app.listen(8080);
+server.listen(8080);
 ```
 
-### .middleware()
+### .process(HttpIncoming)
 
-Connect compatible middleware to execute all parsers in parallel and append the result of each parser
-to `res.locals.podium.context`.
+Metod for processing a incomming http request. It runs all parsers in parallel and append the result of each parser to `HttpIncoming.context`.
```
Comment (alexanbj) [inline README.md]: incoming

File: example/server.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+const { HttpIncoming } = require('@podium/utils');
+const Context = require('../');
+const http = require('http');
+
+// Set up a context with the name 'myLayout'
+const context = new Context({ name: 'myLayout' });
+
+const server = http.createServer(async (req, res) => {
+    // Create a HttpIncomming object
```
Comment (alexanbj) [inline example/server.js]: HttpIncoming

File: README.md [hunk 6]
```diff
@@ -1,10 +1,14 @@
...
 
+Module to generate the context which is passed on requests from a Podium Layout server to a Podium Podlet server.
+
 [![Build Status](https://travis-ci.org/podium-lib/context.svg?branch=master)](https://travis-ci.org/podium-lib/context)
 [![Greenkeeper badge](https://badges.greenkeeper.io/podium-lib/context.svg)](https://greenkeeper.io/)
 [![Known Vulnerabilities](https://snyk.io/test/github/podium-lib/context/badge.svg)](https://snyk.io/test/github/podium-lib/context)
 
-Module to generate, serialize and de-serialize a Podium Context.
+This module is intended for internal use in Podium and is not a module an end user would
+use directly. An end user can though interact with this module through a higher level
```
Comment (digitalsadhu) [inline README.md]: End users will typically interact with this module through higher level modules such as...

File: README.md [hunk 7]
```diff
@@ -14,56 +18,31 @@ $ npm install @podium/context
 
 ## Example
 
-Simple Layout server requesting content from a Podlet server:
+Generate a context which can be passed on to a http request to Podlet:
```
Comment (digitalsadhu) [inline README.md]: to an...
...to a podlet

File: README.md (hunk 2 above)
Comment (digitalsadhu) [reply README.md]: an HttpIncoming

File: README.md [hunk 8]
```diff
@@ -14,56 +18,31 @@ $ npm install @podium/context
...
 
-const app = express();
+    // Run context parsers on the request
+    const incom = await context.process(incoming);
 
-// Attach context de-serialize middleware which will
-// transform the context into a usable object.
-// The context will be stored at: `res.locals.podium.context`
-app.use(Context.deserialize());
+    // Serialize the context into a object that can be passed on
```
Comment (digitalsadhu) [inline README.md]: Serialize the context into an object that can be passed on as HTTP headers on each HTTP request

File: README.md [hunk 9]
```diff
@@ -86,30 +65,30 @@ Each part works as follow:
...
 ### Parsers
 
-A parser works on an inbound request to the Layout server. The parser has access to the
-HTTP `request` and `response` objects in a Layout server. Upon this a parser builds a
-value which will be applied as part of the context which is appended to requests
-to Podlet servers.
+A parser works on an inbound request to the Layout server. A parser is handed a
+[HttpIncoming object](https://github.com/podium-lib/utils/blob/master/lib/http-incoming.js)
+on each request. Upon execution a parser builds a value which will be applied as
+part of the context which is appended to the requests to Podlet servers.
```
Comment (digitalsadhu) [inline README.md]: Parsers operate on inbound requests to the layout server. Each parser is handed an ... for each request. ... context and then appended to all requests made to podlet servers.

File: README.md [hunk 10]
```diff
@@ -86,30 +65,30 @@ Each part works as follow:
...
 
 It's also possible to write custom parsers and append them to the process of constructing
 the context.
 
-### Middleware
+### Processing
 
-The middleware is a Connect compatible middleware which runs on all HTTP requests to
-the Layout server and runs all registered parsers.
+There is a `.process()` method which takes a [HttpIncoming object](https://github.com/podium-lib/utils/blob/master/lib/http-incoming.js)
```
Comment (digitalsadhu) [inline README.md]: ...takes an...

File: README.md [hunk 11]
```diff
@@ -86,30 +65,30 @@ Each part works as follow:
...
+and runs it through all registered parsers in paralell.
```
Comment (digitalsadhu) [inline README.md]: and then runs ... in parallel.

File: README.md [hunk 12]
```diff
@@ -86,30 +65,30 @@ Each part works as follow:
...
 
-The middleware stores the result of each parser as an object on the response at `res.locals.podium.context`.
-This object is "HTTP header like" and can be serialized into headers on an HTTP request
-to a Podlet.
+The result of each parser is stored as an object on `.context` on the passed in
+`HttpIncoming` object. The object stored on `.context` is "HTTP header like" and can
+be serialized into headers on an HTTP request to a Podlet.
```
Comment (digitalsadhu) [inline README.md]: The result of each parser is stored in an object which is set on the `.context` property of the `HttpIncoming` object. This object is "HTTP header like" and can...

File: README.md [hunk 13]
```diff
@@ -192,47 +171,52 @@ This method takes the following arguments:
...
-const express = require('express');
+const { HttpIncoming } = require('@podium/utils');
 const Context = require('@podium/context');
 const Parser = require('my-custom-parser');
+const http = require('http');
 
-const app = express();
-
-const context = new Context('myName');
+// Set up a context and register the cusrom parser
```
Comment (digitalsadhu) [inline README.md]: custom

File: README.md (hunk 4 above)
Comment (digitalsadhu) [inline README.md]: Method ... an incoming HTTP ... appends

File: README.md [hunk 14]
```diff
@@ -192,47 +171,52 @@ This method takes the following arguments:
...
+An instance of a [HttpIncoming object](https://github.com/podium-lib/utils/blob/master/lib/http-incoming.js).
+
 ## Static API
 
 The Context constructor has the following static API:
 
 ### .serialize(headers, context, podletName)
 
-Takes an "HTTP header-ish" object produced by `.middleware()` (the `res.locals.podium.context` object)
+Takes an "HTTP header-ish" object produced by `.process()` (the `HttpIncoming.context` object)
```
Comment (digitalsadhu) [inline README.md]: else where we have used "HTTP header like"

File: README.md [hunk 15]
```diff
@@ -277,10 +265,11 @@ app.get('/', (req, res) => {
 
 ## Internal parsers
 
-This module comes with a set of default parsers which will be applied when `.middleware()` is
+This module comes with a set of default parsers which will be applied when `.process()` is
 run.
 
-Each of these parsers can be configured through with constructor option object by passing an options object to the matching options parameter for the parser in the constructor (see constructor options).
+Each of these parsers can be configured through with constructor option object by passing an options object
+to the matching options parameter for the parser in the constructor (see constructor options).
```
Comment (digitalsadhu) [inline README.md]: ...through the constructor's options object by passing an options object for the specific parser (see constructor options)

**Pull request merged**